src/Dialect/ONNX/ONNXOps.cpp

```diff
@@ -2963,67 +2963,38 @@ LogicalResult ONNXSplitV11Op::inferShapes(
 //===----------------------------------------------------------------------===//
 
 LogicalResult ONNXFlattenOp::verify() {
-
-  if (!hasShapeAndRank(input())) {
-    return success();
-  }
-  auto inTy = input().getType().dyn_cast<ShapedType>();
-  if (!inTy) {
+  if (!hasShapeAndRank(input()))
+    // Won't be able to do any checking at this stage.
     return success();
-  }
 
-  int64_t axisValue = axis();
-  auto inputShape = inTy.getShape();
+  auto inputType = input().getType().cast<ShapedType>();
+  ArrayRef<int64_t> inputShape = inputType.getShape();
   int64_t inputRank = inputShape.size();
+  int64_t axisValue = axis();
 
-  if (axisValue < -1 * inputRank || axisValue > inputRank) {
-    return emitOpError("ONNXFlattenOP: axis() value is out of range");
-  }
+  // axis attribute must be in the range [-r,r], where r = rank(input).
+  if (axisValue < -inputRank || axisValue > inputRank)
+    return onnx_mlir::Diagnostic::attributeOutOfRange(*this->getOperation(),
+        "axis", axisValue,
+        onnx_mlir::Diagnostic::Range<int64_t>(-inputRank, inputRank));
 
   return success();
 }
 
 LogicalResult ONNXFlattenOp::inferShapes(
     std::function<void(mlir::Region &)> doShapeInference) {
-  auto inTy = input().getType().dyn_cast_or_null<RankedTensorType>();
-  if (!inTy) {
+  if (!input().getType().isa<RankedTensorType>())
```

Comment:
I though that this is not the preferred way to test this here, as members are not ranked tensor type.
The code is not very systematic at this time. I believe we have `hasShapeAndRank` and `hasRank` defined in different places that might be better than this. It would be nice to have a consistent approach.

I guess its ok to use a call that does not support memref for shape inference, but not ok for verifiers (as verifier can be called on memrefs as well). Though conistently using the same call may simplify the code.

Reply:
I see many other opportunity to use the same template for the shape inference code of other operators. So I'll do that too.

Reply:
@AlexandreEichenberger done